examples/solvers/find_optimal_parameters.py

```diff
@@ -0,0 +1,234 @@
+    # the parameter to optimize is (lam, sigma) with
+    # sigma the size of the quadratic region
+    #   (parametrizing the moreau envelope of smoothed l1 norm)
+    # lam the regularization strength
+    #   (multiplying the smoothed l1 in the objectiv function)
+    # See lbfgs_tomograhpy_tv.py for more information.
+    def reconstruction(proj_data, parameters):
+        # Extract the separate parameters
+        lam, sigma = parameters
+
+        print('lam = {}, sigma = {}'.format(lam, sigma))
+
+        # We do not allow negative paramters, so return a bogus result
+        if lam <= 0 or sigma <= 0:
+            return np.inf *  space.one()
+
+        # Create data term ||Ax - b||_2^2
+        l2_norm = odl.solvers.L2NormSquared(ray_trafo.range)
+        data_discrepancy = l2_norm * (ray_trafo - proj_data)
+
+        # Create regularizing functional huber(|grad(x)|)
+        gradient = odl.Gradient(space)
+        l1_norm = odl.solvers.GroupL1Norm(gradient.range)
+        smoothed_l1 = odl.solvers.MoreauEnvelope(l1_norm, sigma=sigma)
+        regularizer = smoothed_l1 * gradient
+
+        # Create full objective functional
+        obj_fun = data_discrepancy + lam * regularizer
+
+        # Pick parameters
+        maxiter = 30
+        num_store = 5
+
+        # Run the algorithm
+        x = ray_trafo.domain.zero()
+        odl.solvers.bfgs_method(
+            obj_fun, x, maxiter=maxiter, num_store=num_store,
+            hessinv_estimate=odl.ScalingOperator(
+                    space, 1 / odl.power_method_opnorm(ray_trafo) ** 2))
+
+        return x
+
+    initial_param = [0.1, 0.05]
+
+elif reconstruction_method == 'tv':
+    # Define the reconstruction operator for TV regularized reconstruction
+    # The parameter to optimize is the regularization strength
+    # See chambolle_pock_tomography.py for more information.
+
+    def reconstruction(proj_data, lam):
+        lam = float(lam)
+
+        print('lam = {}'.format(lam))
+
+        # We do not allow negative paramters, so return a bogus result
+        if lam <= 0:
+            return np.inf * space.one()
+
+        # Construct operators and functionals
+        gradient = odl.Gradient(space)
+        op = odl.BroadcastOperator(ray_trafo, gradient)
+
+        g = odl.solvers.ZeroFunctional(op.domain)
+
+        l2_norm = odl.solvers.L2NormSquared(ray_trafo.range).translated(proj_data)
+        l1_norm = lam * odl.solvers.GroupL1Norm(gradient.range)
+        f = odl.solvers.SeparableSum(l2_norm, l1_norm)
+
+        # Select solver parameters
+        op_norm = 1.1 * odl.power_method_opnorm(op)
+
+        # Run the algorithm
+        x = op.domain.zero()
+        odl.solvers.chambolle_pock_solver(
+            x, f, g, op, tau=1.0 / op_norm, sigma=1.0 / op_norm, niter= 200, gamma= 0.3)
+
+        return x
+
+    initial_param = 0.1
+else:
+    raise RuntimeError('unknown reconstruction method')
+
+
+def fom(I0,I1):
+    gradient = odl.Gradient(I0.space)
```

Comment:
Move the definition of the gradient to outside of the function, also rename arguments to something like `x1`, `x2` since this is the standard naming. Also perhaps write a line on what is happening here.